### PR TITLE
Refactors the Light Eater and gives it some flavortext. Some minor behavior changes included.

### PR DESCRIPTION
**Pull request by TemporalOroboros**
## About The Pull Request
- Converts the effects of the nightmares light eater into a component and a couple elements
- Adds some flavor text to the light eater
- Makes the effects of the nightmare's light eater dispel if the armblade is destroyed

## Why It's Good For The Game
- Probably a net increase in code quality
- More flavortext
- The nightmare's light eater effect can be dispelled if you are willing to sacrifice the armblade

## Changelog
:cl:
refactor: The light eater is now a component
balance: The light eater's esoteric effects can be dispelled if you sacrifice the armblade
add: The light eater has some additional flavortext
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
@tgstation/commit-access review and merge please, it's been sitting here for no reason

File: code/datums/components/light_eater.dm [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+/// Handles eating lights
+/datum/component/light_eater/proc/devour(atom/morsel, datum/eater)
+	if(!light_eater_devour(morsel, eater))
+		return FALSE
+
+	morsel.visible_message(
+		"<span class='danger'>Something dark and hungry swarms out of \the [eater] and over \the [morsel]!</span>",
+		"<span class='userdanger'>Something dark and hungry swarms out of \the [eater] and burrows into you!</span>",
+		"<span class='danger'>You can feel a dark hum gnaw at your sight.</span>"
+	)
```
Comment (Rohesie) [inline code/datums/components/light_eater.dm:89]: This might get spammy. You might want to compile this as a list and return it for feedback, perhaps, so a single visible message is sent.

File: code/datums/components/light_eater.dm [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+	if(length(_eaten))
+		eaten_lights = list()
+		var/list/cached_eaten_lights = eaten_lights
+		for(var/food in _eaten)
+			RegisterSignal(food, COMSIG_ATOM_UPDATE_LIGHT, .proc/block_light_update)
+			RegisterSignal(food, COMSIG_PARENT_QDELETING, .proc/deref_eaten_light)
+			RegisterSignal(food, COMSIG_PARENT_EXAMINE, .proc/on_eaten_light_examine)
+			cached_eaten_lights += food
+	return ..()
```
Comment (Rohesie) [inline code/datums/components/light_eater.dm:23]:
```suggestion
/datum/component/light_eater/Initialize(list/_eaten)
	. = ..()
	if(!isatom(parent) && !istype(parent, /datum/reagent))
		return COMPONENT_INCOMPATIBLE

	if(length(_eaten))
		eaten_lights = list()
		var/list/cached_eaten_lights = eaten_lights
		for(var/food in _eaten)
			RegisterSignal(food, COMSIG_ATOM_UPDATE_LIGHT, .proc/block_light_update)
			RegisterSignal(food, COMSIG_PARENT_QDELETING, .proc/deref_eaten_light)
			RegisterSignal(food, COMSIG_PARENT_EXAMINE, .proc/on_eaten_light_examine)
			cached_eaten_lights += food

```
Call parent first on initialization whenever you can.

File: code/__HELPERS/light_eater.dm [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// Procs that are re-used by both the light eater component and element to prevent copypasta and duplicate signals
+
+/// Abstracted as a global proc so I can avoid sending the signal from two places and reduce copypasta
+/proc/light_eater_table_buffet(atom/commisary, datum/light_eater)
```
Comment (Rohesie) [inline code/__HELPERS/light_eater.dm:4]: Could you explain a bit why this proc is not atom-typed, and why you are passing the `light_eaten` (component I assume) as an argument?

File: code/datums/components/light_eater.dm (hunk 1 above)
Comment (TemporalOroboros) [reply code/datums/components/light_eater.dm]: Done

File: code/__HELPERS/light_eater.dm (hunk 3 above)
Comment (TemporalOroboros) [reply code/__HELPERS/light_eater.dm]: That would require remembering what I was thinking a month ago.

File: code/datums/components/light_eater.dm (hunk 2 above)
Comment (TemporalOroboros) [reply code/datums/components/light_eater.dm]: Done

File: code/datums/components/overlay_lighting.dm [hunk 4]
```diff
@@ -135,6 +136,7 @@
 		COMSIG_ATOM_SET_LIGHT_ON,
 		COMSIG_ATOM_SET_LIGHT_FLAGS,
 		COMSIG_ATOM_USED_IN_CRAFT,
+		COMSIG_LIGHT_EATER_ACT,
```
Comment (Rohesie) [inline code/datums/components/overlay_lighting.dm:139]: Are you are registering to `COMSIG_LIGHT_EATER_QUEUE` and unregistering to `COMSIG_LIGHT_EATER_ACT`
This does not seem intended

File: code/datums/elements/light_eater.dm [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+/**
+ * Aggregates a list of the light sources attached to the target atom.
+ *
+ * Arguments:
+ * - [comissary][/atom]: The origin node of all of the light sources to search through.
+ * - [light_eater][/datum]: The light eater being applied to the target.
+ */
+/datum/element/light_eater/proc/table_buffet(atom/commisary, datum/light_eater)
+	. = list()
+	SEND_SIGNAL(commisary, COMSIG_LIGHT_EATER_QUEUE, ., light_eater)
```
Comment (Rohesie) [inline code/datums/elements/light_eater.dm:74]:
```suggestion
	SEND_SIGNAL(commisary, COMSIG_LIGHT_EATER_QUEUE, .)
```
This is already the source of the signal.

File: code/datums/elements/light_eater.dm [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	)
+
+/**
+ * Aggregates a list of the light sources attached to the target atom.
+ *
+ * Arguments:
+ * - [comissary][/atom]: The origin node of all of the light sources to search through.
+ * - [light_eater][/datum]: The light eater being applied to the target.
+ */
+/datum/element/light_eater/proc/table_buffet(atom/commisary, datum/light_eater)
```
Comment (Rohesie) [inline code/datums/elements/light_eater.dm:72]: What is `datum/light_eater`? The component? If so, it shouldn't be passed around.
This doesn't seem to be used, though.

File: code/datums/elements/light_eaten.dm [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+/datum/element/light_eaten/proc/block_light_update(atom/eaten_light)
+	SIGNAL_HANDLER
+	if(eaten_light.light_power <= 0)
+		return NONE
+
+	eaten_light.light_power = 0
+	if(eaten_light.light_range > 0)
+		eaten_light.light_range = 0
+	if(eaten_light.light_on)
+		eaten_light.light_on = FALSE
```
Comment (Rohesie) [inline code/datums/elements/light_eaten.dm:30]: Use the setter procs for these, (like `set_light_power()`), else it will break the light component.

File: code/datums/elements/light_eater.dm (hunk 6 above)
Comment (TemporalOroboros) [reply code/datums/elements/light_eater.dm:72]: It's not the component. I think I made it the parent to the light eater component in case whatever I'm hitting wants to react/change behavior according to that. It's a datum because I made the component compatible with reagents.

Comment (Rohesie) [reply code/datums/elements/light_eater.dm:72]: > 
> 
> It's not the component. I think I made it the parent to the light eater component in case whatever I'm hitting wants to react/change behavior according to that. It's a datum because I made the component compatible with reagents.

Please rephrase, I didn't understand the answer.
What is `datum/light_eater` if not the component? Is this a reference to the element itself, the `src` of the proc? Or to something else?

Comment (TemporalOroboros) [reply code/datums/elements/light_eater.dm:72]: It is the reference to the datum the light eater element is attached to.

File: code/datums/elements/light_eaten.dm (hunk 7 above)
Comment (TemporalOroboros) [reply code/datums/elements/light_eaten.dm:30]: Alright. I guess there's no harm in that.

File: code/datums/components/overlay_lighting.dm (hunk 4 above)
Comment (TemporalOroboros) [reply code/datums/components/overlay_lighting.dm]: Fixed.

**Comment by TemporalOroboros** [comment]
I've done a bit of testing and it works.

File: code/modules/lighting/lighting_atom.dm [hunk 8]
```diff
@@ -112,42 +114,42 @@
...
-
-
+/// Setter for the light power of this atom.
 /atom/proc/set_light_power(new_power)
 	if(new_power == light_power)
 		return
-	SEND_SIGNAL(src, COMSIG_ATOM_SET_LIGHT_POWER, new_power)
 	. = light_power
 	light_power = new_power
+	SEND_SIGNAL(src, COMSIG_ATOM_SET_LIGHT_POWER, new_power, .)
```
Comment (Rohesie) [inline code/modules/lighting/lighting_atom.dm:123]: Why did you move the signal to afterwards?
We try to have them as soon as we can whenever possible, because that allows us to add the option to cancel it, for example, if needed.

File: code/modules/vehicles/mecha/_mecha.dm [hunk 9]
```diff
@@ -1223,3 +1225,15 @@
 		else
 			to_chat(user, "<span class='notice'>None of the equipment on this exosuit can use this ammo!</span>")
 	return FALSE
+
+
+/// Special light eater handling
+/obj/vehicle/sealed/mecha/proc/on_light_eater(obj/vehicle/sealed/mecha, datum/source)
```
Comment (Rohesie) [inline code/modules/vehicles/mecha/_mecha.dm:1231]: The mecha is the source of the signal, not the eater.
Can you fix these names?

File: code/modules/power/lighting.dm [hunk 10]
```diff
@@ -843,6 +845,14 @@
 	sleep(1)
 	qdel(src)
 
+/obj/machinery/light/proc/on_light_eater(obj/machinery/light, datum/light_eater)
+	SIGNAL_HANDLER
+	if(status == 1)
```
Comment (Rohesie) [inline code/modules/power/lighting.dm:850]: What is `status == 1`? Could you fix this magic number?

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 11]
```diff
@@ -111,6 +112,7 @@
 		if(T && istype(radio) && istype(radio.keyslot))
 			radio.keyslot.forceMove(T)
 			radio.keyslot = null
+	UnregisterSignal(src, COMSIG_LIGHT_EATER_ACT)
```
Comment (Rohesie) [inline code/modules/mob/living/silicon/robot/robot.dm:115]: Signals get automatically unregistered when the object is deleted already.

File: code/modules/antagonists/nightmare/nightmare_organs.dm [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+	playsound(user, 'sound/magic/demon_consume.ogg', 50, TRUE)
+
+
+	user.visible_message("<span class='warning'>Blood erupts from [user]'s arm as it reforms into a weapon!</span>", \
+		"<span class='userdanger'>Icy blood pumps through your veins as your arm reforms itself!</span>")
+	user.temporarilyRemoveItemFromInventory(src, TRUE)
+	Insert(user)
+
+/obj/item/organ/heart/nightmare/Insert(mob/living/carbon/M, special = 0)
+	..()
```
Comment (Rohesie) [inline code/modules/antagonists/nightmare/nightmare_organs.dm:58]:
```suggestion
	. = ..()
```

File: code/modules/antagonists/nightmare/nightmare_organs.dm [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+	if(special != HEART_SPECIAL_SHADOWIFY)
+		blade = new/obj/item/light_eater
+		M.put_in_hands(blade)
+
+/obj/item/organ/heart/nightmare/Remove(mob/living/carbon/M, special = 0)
+	respawn_progress = 0
+	if(blade && special != HEART_SPECIAL_SHADOWIFY)
+		M.visible_message("<span class='warning'>\The [blade] disintegrates!</span>")
+		QDEL_NULL(blade)
+	..()
```
Comment (Rohesie) [inline code/modules/antagonists/nightmare/nightmare_organs.dm:68]:
```suggestion
	return ..()
```

File: code/modules/antagonists/nightmare/nightmare_organs.dm [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+		"<span class='danger'>[src] feels unnaturally cold in your hands. You raise [src] your mouth and devour it!</span>")
+	playsound(user, 'sound/magic/demon_consume.ogg', 50, TRUE)
+
+
+	user.visible_message("<span class='warning'>Blood erupts from [user]'s arm as it reforms into a weapon!</span>", \
+		"<span class='userdanger'>Icy blood pumps through your veins as your arm reforms itself!</span>")
+	user.temporarilyRemoveItemFromInventory(src, TRUE)
+	Insert(user)
+
+/obj/item/organ/heart/nightmare/Insert(mob/living/carbon/M, special = 0)
```
Comment (Rohesie) [inline code/modules/antagonists/nightmare/nightmare_organs.dm:57]: What does `special = 0` mean?

File: code/modules/antagonists/nightmare/nightmare_organs.dm [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+/obj/item/organ/heart/nightmare
+	name = "heart of darkness"
+	desc = "An alien organ that twists and writhes when exposed to light."
+	icon = 'icons/obj/surgery.dmi'
+	icon_state = "demon_heart-on"
+	color = "#1C1C1C"
+	var/respawn_progress = 0
+	var/obj/item/light_eater/blade
+	decay_factor = 0
```
Comment (Rohesie) [inline code/modules/antagonists/nightmare/nightmare_organs.dm:37]: Can you put redefined variables over the newly-defined ones, and document new variables?

File: code/modules/antagonists/nightmare/nightmare_organs.dm [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+		visible_message("<span class='warning'>[M] thrashes as [src] takes root in [M.p_their()] body!</span>")
+	var/obj/effect/proc_holder/spell/targeted/shadowwalk/SW = new
+	M.AddSpell(SW)
+	shadowwalk = SW
+
+
+/obj/item/organ/brain/nightmare/Remove(mob/living/carbon/M, special = 0)
+	if(shadowwalk)
+		M.RemoveSpell(shadowwalk)
+	..()
```
Comment (Rohesie) [inline code/modules/antagonists/nightmare/nightmare_organs.dm:26]:
```suggestion
	return ..()
```

File: code/modules/antagonists/nightmare/nightmare_organs.dm [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+
+/obj/item/organ/brain/nightmare
+	name = "tumorous mass"
+	desc = "A fleshy growth that was dug out of the skull of a Nightmare."
+	icon_state = "brain-x-d"
+	var/obj/effect/proc_holder/spell/targeted/shadowwalk/shadowwalk
+
+/obj/item/organ/brain/nightmare/Insert(mob/living/carbon/M, special = 0)
+	..()
```
Comment (Rohesie) [inline code/modules/antagonists/nightmare/nightmare_organs.dm:14]:
```suggestion
	. = ..()
```

File: code/modules/antagonists/nightmare/nightmare.dm [hunk 18]
```diff
@@ -3,3 +3,10 @@
 	show_in_antagpanel = FALSE
 	show_name_in_check_antagonists = TRUE
 	show_to_ghosts = TRUE
+
+
+
+
+
+
+
```
Comment (Rohesie) [inline code/modules/antagonists/nightmare/nightmare.dm:12]: What's this about?

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 19]
```diff
@@ -1106,6 +1116,7 @@ GLOBAL_LIST_EMPTY(PDAs)
 
 /obj/item/pda/Destroy()
 	GLOB.PDAs -= src
+	UnregisterSignal(src, COMSIG_LIGHT_EATER_ACT)
```
Comment (Rohesie) [inline code/game/objects/items/devices/PDA/PDA.dm:1119]: No need to unregister source signals on destroy, that's done automatically already.

**Review by Rohesie — CHANGES_REQUESTED** [review]

File: code/modules/lighting/lighting_atom.dm (hunk 8 above)
Comment (TemporalOroboros) [reply code/modules/lighting/lighting_atom.dm:123]: Because this is the standard for setter procs with signals as far as I can tell.

File: code/modules/power/lighting.dm (hunk 10 above)
Comment (TemporalOroboros) [reply code/modules/power/lighting.dm:850]: Sure. I think I just copied this from the old proc.

File: code/modules/antagonists/nightmare/nightmare_organs.dm (hunk 14 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare_organs.dm:57]: I'm pretty sure this just means that if it replaces a vital organ it kills by default. It was like that when I moved it.

File: code/modules/antagonists/nightmare/nightmare_organs.dm (hunk 15 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare_organs.dm:37]: Sure.

File: code/modules/antagonists/nightmare/nightmare_organs.dm (hunk 16 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare_organs.dm:26]: It was like that originally and I don't want to break something by changing it. I'll take a look.

File: code/modules/antagonists/nightmare/nightmare_organs.dm (hunk 17 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare_organs.dm:14]: Same

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 19 above)
Comment (TemporalOroboros) [reply code/game/objects/items/devices/PDA/PDA.dm:1119]: It's the standard for components and the like.

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 11 above)
Comment (TemporalOroboros) [reply code/modules/mob/living/silicon/robot/robot.dm:115]: I know. But this is the standard format.

File: code/modules/vehicles/mecha/_mecha.dm (hunk 9 above)
Comment (TemporalOroboros) [reply code/modules/vehicles/mecha/_mecha.dm:1231]: Sure.

Comment (TemporalOroboros) [reply code/modules/vehicles/mecha/_mecha.dm:1231]: Done.

File: code/modules/power/lighting.dm (hunk 10 above)
Comment (TemporalOroboros) [reply code/modules/power/lighting.dm:850]: Done

File: code/modules/antagonists/nightmare/nightmare_organs.dm (hunk 15 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare_organs.dm:37]: Done.

File: code/modules/antagonists/nightmare/nightmare_organs.dm (hunk 16 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare_organs.dm:26]: Done

File: code/modules/antagonists/nightmare/nightmare_organs.dm (hunk 17 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare_organs.dm:14]: Done

File: code/modules/antagonists/nightmare/nightmare.dm (hunk 18 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare.dm:12]: I don't know, but I removed it.

File: code/modules/antagonists/nightmare/nightmare_organs.dm (hunk 13 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare_organs.dm:68]: Done

File: code/modules/antagonists/nightmare/nightmare_organs.dm (hunk 14 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare_organs.dm:57]: I've replaced it with FALSE.

File: code/modules/antagonists/nightmare/nightmare_organs.dm (hunk 12 above)
Comment (TemporalOroboros) [reply code/modules/antagonists/nightmare/nightmare_organs.dm:58]: Done

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 19 above)
Comment (Rohesie) [reply code/game/objects/items/devices/PDA/PDA.dm:1119]: Check the definition of `Destroy()` at the base.
This is already handled by default.

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 11 above)
Comment (Rohesie) [reply code/modules/mob/living/silicon/robot/robot.dm:115]: Where is this standard taken from?

File: code/modules/lighting/lighting_atom.dm (hunk 8 above)
Comment (Rohesie) [reply code/modules/lighting/lighting_atom.dm:123]: This is not the standard at all.

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 19 above)
Comment (TemporalOroboros) [reply code/game/objects/items/devices/PDA/PDA.dm:1119]: Done

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 11 above)
Comment (TemporalOroboros) [reply code/modules/mob/living/silicon/robot/robot.dm:115]: Done. More overhead, but done.

File: code/modules/lighting/lighting_atom.dm (hunk 8 above)
Comment (TemporalOroboros) [reply code/modules/lighting/lighting_atom.dm:123]: Done. More overhead, but done.

**Comment by TemporalOroboros** [comment]
I tested this and it works. I'm still waiting for ninja to get back to me on the unregistering signals bit.

**Comment by Rohesie** [comment]
Ping him on discord if you want an answer.

**Comment by TemporalOroboros** [comment]
Done and tested.

File: code/__DEFINES/dcs/signals.dm [hunk 20]
```diff
@@ -219,16 +217,32 @@
...
-///Called right before the atom changes the value of light_power to a different one, from base atom/set_light_power(): (new_power)
+// Lighting:
+///from base of [atom/proc/set_light]: (l_range, l_power, l_color, l_on)
+#define COMSIG_ATOM_SET_LIGHT "atom_set_light"
+	/// Blocks [/atom/proc/set_light], [/atom/proc/set_light_power], [/atom/proc/set_light_range], [/atom/proc/set_light_color], [/atom/proc/set_light_on], and [/atom/proc/set_light_flags].
+	#define COMPONENT_BLOCK_LIGHT_UPDATE	(1<<0)
+///Called right before the atom changes the value of light_power to a different one, from base [atom/proc/set_light_power]: (new_power, old_power)
 #define COMSIG_ATOM_SET_LIGHT_POWER "atom_set_light_power"
-///Called right before the atom changes the value of light_color to a different one, from base atom/set_light_color(): (new_color)
+///Called right after the atom changes the value of light_power to a different one, from base of [/atom/proc/set_light_power]: (new_power, old_power)
```
Comment (Rohesie) [inline code/__DEFINES/dcs/signals.dm:227]: Fix this please. Make the signal be sent before, not after.

**Review by Rohesie — CHANGES_REQUESTED** [review]

File: code/__DEFINES/dcs/signals.dm (hunk 20 above)
Comment (TemporalOroboros) [reply code/__DEFINES/dcs/signals.dm:227]: I did. You told me to make two signals so I made two signals.

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/datums/elements/light_eater.dm [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+			UnregisterSignal(source, COMSIG_MOVABLE_IMPACT)
+			if(isitem(source))
+				UnregisterSignal(source, list(
+					COMSIG_ITEM_AFTERATTACK,
+					COMSIG_ITEM_HIT_REACT,
+				))
+			else if(isprojectile(source))
+				UnregisterSignal(source, COMSIG_PROJECTILE_ON_HIT)
+	else if(istype(source, /datum/reagent))
+		UnregisterSignal(source, COMSIG_REAGENT_EXPOSE_ATOM)
```
[the remark this answers is not in the document]
Comment (TemporalOroboros) [reply code/datums/elements/light_eater.dm]: Done

**Comment by TemporalOroboros** [comment]
Alright. I did it. Also tested.

File: code/__DEFINES/dcs/signals.dm (hunk 20 above)
Comment (Rohesie) [reply code/__DEFINES/dcs/signals.dm:229]: Right, I see that now.
```suggestion
///Called right after the atom changes the value of light_power to a different one, from base of [/atom/proc/set_light_power]: (old_power)
```
No need to pass the new power as an argument, since the source is already being passed.

**Review by Rohesie — CHANGES_REQUESTED** [review]
Merge conflict and one change request.
Should be ready for merge after.

File: code/__DEFINES/dcs/signals.dm (hunk 20 above)
Comment (TemporalOroboros) [reply code/__DEFINES/dcs/signals.dm:229]: Done

**Comment by TemporalOroboros** [comment]
Done and tested.

File: code/__DEFINES/dcs/signals.dm [hunk 22]
```diff
@@ -221,16 +219,32 @@
...
+	/// Blocks [/atom/proc/set_light], [/atom/proc/set_light_power], [/atom/proc/set_light_range], [/atom/proc/set_light_color], [/atom/proc/set_light_on], and [/atom/proc/set_light_flags].
+	#define COMPONENT_BLOCK_LIGHT_UPDATE (1<<0)
+///Called right before the atom changes the value of light_power to a different one, from base [atom/proc/set_light_power]: (new_power)
 #define COMSIG_ATOM_SET_LIGHT_POWER "atom_set_light_power"
-///Called right before the atom changes the value of light_color to a different one, from base atom/set_light_color(): (new_color)
+///Called right after the atom changes the value of light_power to a different one, from base of [/atom/proc/set_light_power]: (new_power, old_power)
+#define COMSIG_ATOM_UPDATE_LIGHT_POWER "atom_update_light_power"
+///Called right before the atom changes the value of light_range to a different one, from base [atom/proc/set_light_range]: (new_range)
+#define COMSIG_ATOM_SET_LIGHT_RANGE "atom_set_light_range"
+///Called right after the atom changes the value of light_range to a different one, from base of [/atom/proc/set_light_range]: (new_range, old_range)
```
Comment (Rohesie) [inline code/__DEFINES/dcs/signals.dm:233]: Sorry if I wasn't clear, but the same logic applies to all the new signals.
Send only the old value forward and update the comment.
The new value is already known since we're passing the source.

File: code/__DEFINES/dcs/signals.dm (hunk 20 above)
Comment (Rohesie) [reply code/__DEFINES/dcs/signals.dm:229]: Update the comment as well, and apply this to all the new signals, not just this one.

**Review by Rohesie — CHANGES_REQUESTED** [review]

File: code/__DEFINES/dcs/signals.dm (hunk 22 above)
Comment (TemporalOroboros) [reply code/__DEFINES/dcs/signals.dm]: Done

File: code/datums/components/overlay_lighting.dm [hunk 23]
```diff
@@ -107,12 +107,13 @@
...
-	RegisterSignal(parent, COMSIG_ATOM_SET_LIGHT_RANGE, .proc/set_range)
-	RegisterSignal(parent, COMSIG_ATOM_SET_LIGHT_POWER, .proc/set_power)
-	RegisterSignal(parent, COMSIG_ATOM_SET_LIGHT_COLOR, .proc/set_color)
-	RegisterSignal(parent, COMSIG_ATOM_SET_LIGHT_ON, .proc/on_toggle)
-	RegisterSignal(parent, COMSIG_ATOM_SET_LIGHT_FLAGS, .proc/on_light_flags_change)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_RANGE, .proc/set_range)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_POWER, .proc/set_power)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_COLOR, .proc/set_color)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_ON, .proc/on_toggle)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_FLAGS, .proc/on_light_flags_change)
```
Comment (Rohesie) [inline code/datums/components/overlay_lighting.dm:114]: Why are these changed? They worked fine with the pre-signal.
Revert this change, please.

**Review by Rohesie — CHANGES_REQUESTED** [review]

File: code/datums/components/overlay_lighting.dm (hunk 23 above)
Comment (TemporalOroboros) [reply code/datums/components/overlay_lighting.dm:114]: Because they now go after the setter.
